server/api/diagnose.go

```diff
@@ -0,0 +1,156 @@
+		memberLostPeers:             Recommended{modMember, levelMajor, "some PD instances is down", "please check host load and traffic"},
+		memberLostPeersMoreThanHalf: Recommended{modMember, levelCritical, "more than half PD instances is down", "please check host load and traffic"},
+		memberLeaderChanged:         Recommended{modMember, levelMinor, "PD cluster leader is changed", "please check host load and traffic"},
+		tikvCap70:                   Recommended{modTiKV, levelWarning, "some TiKV stroage used more than 70%", "plase add TiKV node"},
+		tikvCap80:                   Recommended{modTiKV, levelMinor, "some TiKV stroage used more than 80%", "plase add TiKV node"},
+		tikvCap90:                   Recommended{modTiKV, levelMajor, "some TiKV stroage used more than 90%", "plase add TiKV node"},
+		tikvLostPeers:               Recommended{modTiKV, levelWarning, "some TiKV lost connect", "plase check network"},
+		tikvLostPeersLongTime:       Recommended{modTiKV, levelMajor, "some TiKV lost connect more than 1h", "plase check network"},
+	}
+)
+
+type diagnoseHandler struct {
+	svr *server.Server
+	rd  *render.Render
+}
+
+func newDiagnoseHandler(svr *server.Server, rd *render.Render) *diagnoseHandler {
+	return &diagnoseHandler{
+		svr: svr,
+		rd:  rd,
+	}
+}
+
+func diagnosePD(key diagnoseType) *Recommended {
+	d, ok := diagnoseMap[key]
+	if !ok {
+		return &Recommended{}
+	}
+	return &d
+}
+
+func (d *diagnoseHandler) membersDiagnose(rdd *[]*Recommended) error {
+	var lenMembers, lostMembers int
+	var changedLeaderLess1Min bool
+	req := &pdpb.GetMembersRequest{Header: &pdpb.RequestHeader{ClusterId: d.svr.ClusterID()}}
+	members, err := d.svr.GetMembers(context.Background(), req)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	lenMembers = len(members.Members)
+	if lenMembers > 0 {
+		for _, m := range members.Members {
+			pm, err := getEtcdPeerStats(m.ClientUrls[0])
+			if err != nil {
+				//get peer etcd failed
+				lostMembers++
+				continue
+			}
+			if time.Now().Sub(pm.LeaderInfo.StartTime) < time.Duration(3*time.Minute) {
+				changedLeaderLess1Min = true
+			}
+		}
+	} else {
+		return errors.Errorf("get PD member error")
+	}
+	if changedLeaderLess1Min {
+		*rdd = append(*rdd, diagnosePD(memberLeaderChanged))
+	}
+	if lenMembers-lostMembers == 1 {
+		// only one pd running
+		*rdd = append(*rdd, diagnosePD(memberOneInstance))
+	}
+	if lostMembers > 0 {
+		// some pd can not connect
+		*rdd = append(*rdd, diagnosePD(memberLostPeers))
+	}
+	if (lenMembers-lostMembers)%2 == 0 {
+		// alived pd numbers is even
+		*rdd = append(*rdd, diagnosePD(memberEvenInstance))
+	}
+	if float64(lenMembers)/2 < float64(lostMembers) {
+		*rdd = append(*rdd, diagnosePD(memberLostPeersMoreThanHalf))
+	}
+	return nil
+}
+
+func (d *diagnoseHandler) tikvDiagnose(rdd *[]*Recommended) error {
+	return nil
+}
+
+func (d *diagnoseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	rdd := []*Recommended{}
```

Comment:
I prefer `membersDiagnose()` return `rdd`, there is no need to define `rdd` here.

Reply:
only `membersDiagnose` now,
will add other functions (like `tikvDiagnose`)